### PR TITLE
build: produce bundle indexes for `@angular` modules

### PR DESCRIPTION
**Pull request by chuckjaz**
**What kind of change does this PR introduce?** (check one with "x")
```
[x] Feature
[x] Build related changes
```

**What is the current behavior?** (You can also link to an open issue here)

The bundles produced by the angular build are not usable by an AOT application because the compiler produces deep link imports into the Angular bundles.

**What is the new behavior?**

Angular bundles are now usable by AOT applications. The private symbols needed by the compiler are exported as shrouded names that begin with the latin barred o, 'ɵ'. 

**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: modules/@angular/compiler/src/aot/static_symbol_resolver.ts [hunk 1]
```diff
@@ -202,9 +202,16 @@ export class StaticSymbolResolver {
           new Set<string>(Object.keys(metadata['metadata']).map(unescapeIdentifier));
       Object.keys(metadata['metadata']).forEach((metadataKey) => {
         const symbolMeta = metadata['metadata'][metadataKey];
-        resolvedSymbols.push(this.createResolvedSymbol(
-            this.getStaticSymbol(filePath, unescapeIdentifier(metadataKey)), topLevelSymbolNames,
-            symbolMeta));
+        const name = unescapeIdentifier(metadataKey);
+        const canonicalSymbol = this.getStaticSymbol(filePath, name);
+        if (metadata['importAs']) {
```
Comment (tbosch) [inline modules/@angular/compiler/src/aot/static_symbol_resolver.ts]: Do we still need this?

File: modules/@angular/compiler/test/aot/compiler_spec.ts [hunk 2]
```diff
@@ -70,6 +73,53 @@ describe('compiler', () => {
...
+      aotHost = new MockAotCompilerHost(host);
+    });
+
+    // Restore reflector since AoT compiler will update it with a new static reflector
+    afterEach(() => { reflector.updateCapabilities(new ReflectionCapabilities()); });
+
+    it('should compile',
+       async(() => compile(host, aotHost, expectNoDiagnostics).then(generatedFiles => {
+         expect(generatedFiles.find(f => /app\.component\.ngfactory\.ts/.test(f.genFileUrl)))
+             .not.toBeUndefined();
```
Comment (tbosch) [inline modules/@angular/compiler/test/aot/compiler_spec.ts]: double negation. Maybe `.toBeDefined()`?

File: tools/@angular/tsc-wrapped/src/bundler.ts [hunk 3]
```diff
@@ -0,0 +1,551 @@
...
+  // Produced by indirectly by exportAll() for symbols are re-exported by another symbol.
+  reexportedAs?: Symbol;
+
+  // Produced by canonicalizeSymbols() for all symbols. A symbol is private if it is not
+  // exported by the index.
+  isPrivate?: boolean;
+
+  // Produced by canonicalizeSymbols() for all symbols. This it the symbol that should be
+  // imported by factories. For private symbols this is the declaration symbol.
+  canonicalSymbol?: Symbol;
```
Comment (tbosch) [inline tools/@angular/tsc-wrapped/src/bundler.ts]: maybe a better name?

File: tools/@angular/tsc-wrapped/src/bundler.ts [hunk 4]
```diff
@@ -0,0 +1,551 @@
...
+
+  // Produced by canonicalizeSymbols() for all symbols. This the symbol that originally
+  // declared the value and should be used to fetch the value.
+  declaration?: Symbol;
+
+  // A symbol is referenced if it is exported from index or referenced by the value of
+  // a referenced symbol's value.
+  referenced?: boolean;
+
+  // Only valid for referenced symbols. Produces by convertSymbols().
```
Comment (tbosch) [inline tools/@angular/tsc-wrapped/src/bundler.ts]: for `declared` symbols

File: tools/@angular/tsc-wrapped/src/bundler.ts (hunk 3 above)
Comment (tbosch) [inline tools/@angular/tsc-wrapped/src/bundler.ts]: maybe have a `rootExportSymbol` here, and have a top level function `canonicalSymbol` as a function.

Comment (tbosch) [inline tools/@angular/tsc-wrapped/src/bundler.ts]: Add comment what `canonical` means...

**Review by tbosch — APPROVED** [review]
LGTM for compiler and tsc-wrapped.

File: modules/@angular/compiler/src/aot/static_symbol_resolver.ts (hunk 1 above)
Comment (chuckjaz) [inline modules/@angular/compiler/src/aot/static_symbol_resolver.ts]: Yes, as discussed offline. This is the module name the user declared in the tsconfig.json file that the compiler should generate for references to symbols in a bundle.

File: modules/@angular/compiler/test/aot/compiler_spec.ts (hunk 2 above)
Comment (chuckjaz) [inline modules/@angular/compiler/test/aot/compiler_spec.ts]: done.

File: tools/@angular/tsc-wrapped/src/bundler.ts (hunk 3 above)
Comment (chuckjaz) [inline tools/@angular/tsc-wrapped/src/bundler.ts]: Improved the comment.

File: tools/@angular/tsc-wrapped/src/bundler.ts (hunk 4 above)
Comment (chuckjaz) [inline tools/@angular/tsc-wrapped/src/bundler.ts]: Actually canonical symbols. Updated the comment.

File: modules/@angular/compiler/test/aot/compiler_spec.ts [hunk 5]
```diff
@@ -55,17 +58,64 @@ describe('compiler', () => {
...
+             .toBeDefined();
+       })));
+  });
+});
+
+describe('compiler (bundled Angular)', () => {
+  let angularFiles: Map<string, string>;
+
+  beforeAll(() => {
+    const emittingHost = new EmittingCompilerHost(['@angular/core/index'], {emitMetadata: false});
```
Comment (IgorMinar) [inline modules/@angular/compiler/test/aot/compiler_spec.ts]: is `/index` right?

File: modules/@angular/upgrade/static.ts [hunk 6]
```diff
@@ -12,10 +12,4 @@
...
  * Angular 1 and Angular 2+ to run side by side in the same application.
  */
-export {downgradeComponent} from './src/common/downgrade_component';
-export {downgradeInjectable} from './src/common/downgrade_injectable';
-export {VERSION} from './src/common/version';
-export {UpgradeComponent} from './src/static/upgrade_component';
-export {UpgradeModule} from './src/static/upgrade_module';
-
-// This file only re-exports content of the `src` folder. Keep it that way.
+export * from './public_static';
```
Comment (IgorMinar) [inline modules/@angular/upgrade/static.ts]: is this right? I thought that this file was supposed to be generated.

File: modules/@angular/upgrade/public_static.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
+/**
```
Comment (IgorMinar) [inline modules/@angular/upgrade/public_static.ts]: can you rename this file to public_api_static.ts for consistency?

File: tools/public_api_guard/common/index.d.ts [hunk 8]
```diff
@@ -10,6 +10,10 @@ export declare class AsyncPipe implements OnDestroy, PipeTransform {
     transform<T>(obj: Observable<T>): T | null;
 }
 
+/** @deprecated */
+export declare class CommonDeprecatedModule {
```
Comment (IgorMinar) [inline tools/public_api_guard/common/index.d.ts]: why is this export added now?

File: tools/public_api_guard/router/index.d.ts [hunk 9]
```diff
@@ -345,6 +345,9 @@ export declare class RouterStateSnapshot extends Tree<ActivatedRouteSnapshot> {
 /** @stable */
 export declare type Routes = Route[];
 
+/** @experimental */
+export declare const ROUTES: InjectionToken<Route[][]>;
```
Comment (IgorMinar) [inline tools/public_api_guard/router/index.d.ts]: why is this export added now?

File: tools/@angular/tsc-wrapped/src/options.ts [hunk 10]
```diff
@@ -25,6 +25,41 @@ interface Options extends ts.CompilerOptions {
...
+  // supplied in the `files` field. This must be of the form found in a import
+  // declaration. For example, if the library index is in `index.ts` then the
+  // `libraryIndex` field should be `"./index"`.
+  libraryIndex?: string;
+
+  // Preferred module name to use for importing the generated bundle. References
+  // generated by `ngc` will use this module name when importing symbols from the
+  // generated bundle. This is only meaningful when `bundleIndex` is also supplied. It is
+  // otherwise ignored.
+  importAs?: string;
```
Comment (IgorMinar) [inline tools/@angular/tsc-wrapped/src/options.ts]: I think that all three of these names are non-intuitive. Maybe we can reconsider these as a follow up change before rc.0.

**Review by IgorMinar — CHANGES_REQUESTED** [review]

File: modules/@angular/upgrade/static.ts (hunk 6 above)
Comment (chuckjaz) [inline modules/@angular/upgrade/static.ts]: Agreed to change this to public_api_static and to add a comment to this file and similar files that indicate it is only for use in the editor and the global compile.

File: tools/public_api_guard/common/index.d.ts (hunk 8 above)
Comment (chuckjaz) [inline tools/public_api_guard/common/index.d.ts]: All components need to be in a module or an error is generated by the compiler. With this change, the module must also be public exported from bundle index or its metadata will be lost and the compiler will also generate the error.

Agreed offline to change the name to `DeprecatedCommonModule`

File: tools/public_api_guard/router/index.d.ts (hunk 9 above)
Comment (chuckjaz) [inline tools/public_api_guard/router/index.d.ts]: This is necessary to have for the lazy route calculation in angular cli. It must be public now to make its reference have a stable name that can be found by angular cli. Otherwise, will be given a shrouded name that is not predictable.

Also, ui-router needs to be able to import this symbol so the lazy route calculation will also work with ui-router. ui-router currently uses a deep import for this symbol which will no longer work deep imported symbols are no longer treaded as equivalent to the  private symbol. They can change their import to be `@angular/router`.

File: tools/@angular/tsc-wrapped/src/options.ts (hunk 10 above)
Comment (chuckjaz) [inline tools/@angular/tsc-wrapped/src/options.ts]: Maybe `bundleIndex: { name: string, publicApi?: string, importAs?: string }`.

File: modules/@angular/compiler/test/aot/compiler_spec.ts (hunk 5 above)
Comment (chuckjaz) [inline modules/@angular/compiler/test/aot/compiler_spec.ts]: Yes. This is unrelated to the changes we needed to make for tsickle

File: tools/@angular/tsc-wrapped/src/options.ts (hunk 10 above)
Comment (IgorMinar) [inline tools/@angular/tsc-wrapped/src/options.ts]: how about `metadataBundle: { outFile: string, file?: string, importAs?: string }` - to align this with existing property names in tsconfig.json?

I'm fine with this being a follow up change before rc.0. let's discuss this one tomorrow.

**Review by IgorMinar — APPROVED** [review]
